nano/node/socket.hpp

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <boost/asio.hpp>
+#include <boost/optional.hpp>
+#include <chrono>
+#include <deque>
+#include <memory>
+#include <utility>
+#include <vector>
+
+namespace nano
+{
+class node;
+class server_socket;
+
+/** Socket class for tcp clients and newly accepted connections */
+class socket : public std::enable_shared_from_this<nano::socket>
```

Comment:
Socket needs to have a virtual destructor or its children won't clean up correctly. Is it a problem to call close_internal within ~socket or should we be checking to make sure the socket isn't deconstructed outside a specific execution context?

Reply:
Fixed virtual dtor. Calling from d'tor should be fine, but will verify.